### PR TITLE
feat: group filter

### PR DESCRIPTION
**Pull request by yury-s**

### CODE REVIEW

File: packages/playwright-test/src/runner.ts [hunk 1]
```diff
@@ -1023,6 +1047,12 @@ class ListModeReporter implements Reporter {
   }
 }
 
+function fileMatcherFrom(testFileFilters?: TestFileFilter[]): Matcher {
+  if (testFileFilters?.length)
+    return  createFileMatcher(testFileFilters.map(e => e.re || e.exact || ''))
```
Comment (dgozman) [inline packages/playwright-test/src/runner.ts:1052]: nit: styling is off!

File: packages/playwright-test/src/runner.ts [hunk 2]
```diff
@@ -258,71 +264,90 @@ export class Runner {
...
+      testTitleMatcher
+    }));
+  }
+
+  private _collectProjects(projectNames: string[]): FullProjectInternal[] {
+    const projectsToFind = new Set<string>();
+    const unknownProjects = new Map<string, string>();
+    projectNames.forEach(n => {
+      const name = n.toLocaleLowerCase();
+      projectsToFind!.add(name);
```
Comment (dgozman) [inline packages/playwright-test/src/runner.ts:320]: nit: don't need exclamation marks

File: packages/playwright-test/src/runner.ts [hunk 3]
```diff
@@ -258,71 +264,90 @@ export class Runner {
...
       const unknownProjectNames = Array.from(unknownProjects.values()).map(n => `"${n}"`).join(', ');
       throw new Error(`Project(s) ${unknownProjectNames} not found. Available named projects: ${names.map(name => `"${name}"`).join(', ')}`);
     }
+    return projects;
+  }
 
+  private async _collectFiles(runPhase: RunPhase): Promise<Map<FullProjectInternal, string[]>> {
+    const projectNames = runPhase.map(p => p.projectName);
+    const projects = this._collectProjects(projectNames);
+    const projectToGroupEntry = new Map(runPhase.map(p => [p.projectName.toLocaleLowerCase(), p]));
```
Comment (dgozman) [inline packages/playwright-test/src/runner.ts:342]: Add config validation for new fields.

File: packages/playwright-test/src/runner.ts [hunk 4]
```diff
@@ -385,6 +411,7 @@ export class Runner {
       for (const [project, files] of filesByProject) {
         const grepMatcher = createTitleMatcher(project.grep);
         const grepInvertMatcher = project.grepInvert ? createTitleMatcher(project.grepInvert) : null;
+        const groupTitleMatcher = phase.find(p => p.projectName.toLocaleLowerCase() === project.name.toLocaleLowerCase())!.testTitleMatcher;
```
Comment (dgozman) [inline packages/playwright-test/src/runner.ts:414]: Let's explore having a class for RunPhase or something like that?

File: packages/playwright-test/src/runner.ts [hunk 5]
```diff
@@ -258,71 +264,90 @@ export class Runner {
...
+              phase.push({
+                projectName: p,
+                testFileMatcher: () => true,
+                testTitleMatcher: () => true,
+              });
+            } else {
+              const testMatch = p.testMatch ? createFileMatcher(p.testMatch) : () => true;
+              const testIgnore = p.testIgnore ? createFileMatcher(p.testIgnore) : undefined;
+              const grep = p.grep ? createTitleMatcher(p.grep) : () => true;
+              const grepInvert = p.grepInvert ? createTitleMatcher(p.grepInvert) : undefined;
```
Comment (dgozman) [inline packages/playwright-test/src/runner.ts:287]:
```suggestion
              const grepInvert = p.grepInvert ? createTitleMatcher(p.grepInvert) : () => false;
```

File: packages/playwright-test/src/runner.ts [hunk 6]
```diff
@@ -258,71 +264,90 @@ export class Runner {
...
-              projectFilter.push(...p.project);
+            if (isString(p)) {
+              phase.push({
+                projectName: p,
+                testFileMatcher: () => true,
+                testTitleMatcher: () => true,
+              });
+            } else {
+              const testMatch = p.testMatch ? createFileMatcher(p.testMatch) : () => true;
+              const testIgnore = p.testIgnore ? createFileMatcher(p.testIgnore) : undefined;
```
Comment (dgozman) [inline packages/playwright-test/src/runner.ts:285]:
```suggestion
              const testIgnore = p.testIgnore ? createFileMatcher(p.testIgnore) : () => false;
```

**Review by dgozman — APPROVED** [review]

File: packages/playwright-test/src/runner.ts (hunk 4 above)
Comment (yury-s) [reply packages/playwright-test/src/runner.ts:414]: I'll do it in a separate PR as discussed offline.

**Pull request merged**